File: 0767-reorganize-string/0767-reorganize-string.py
# TC- O(nlogk)
# SC- O(1)

class Solution:
    def reorganizeString(self, s: str) -> str:
        # To avoid adjacent duplicates, always place the character with the highest remaining frequency, but make sure we don’t use the same character twice in a row.
        
        # Push the chars into a max_heap by their freq of occ
        max_heap = []
        freq_map = Counter(s) 
        for ch, freq in freq_map.items():
            heapq.heappush(max_heap, (-freq, ch))
        
        result = []
        prev = None
        while max_heap:
            # pop top element
            freq, ch = heapq.heappop(max_heap)

            result.append(ch)

            if prev and prev[0]!=0:
                heapq.heappush(max_heap, prev)

            # decrement it and if not 0, push it back
            freq += 1
            prev = (freq, ch)

        res = "".join(result)
        return res if len(res)==len(s) else ""

File: 0767-reorganize-string/0767-reorganize-string.cpp
// TC- O(nlog26)
// SC- O(26)

class Solution {
public:
    string reorganizeString(string s) {
        unordered_map<char, int> freq_map;
        for(auto& ch:s)
            freq_map[ch]++;

        priority_queue<pair<int, char>> max_heap;
        for(auto& [ch, freq]:freq_map)
            max_heap.push({freq, ch});
        
        pair<int, char> prev = {0, '#'};
        string ans;

        while(!max_heap.empty()){
            auto curr = max_heap.top();
            max_heap.pop();

            ans += curr.second;

            if (prev.first>0)
                max_heap.push(prev);

            --curr.first;
            prev = curr;
        }

        return ans.size()==s.size() ? ans:"";
    }
};